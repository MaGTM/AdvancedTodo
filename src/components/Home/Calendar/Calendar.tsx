import React, {FC} from 'react';
import {Box, Grid, Typography} from "@mui/material";
import {CalendarItem} from '../../../styledComponents/CalendarItem';
import {ICalendar, ICalendarState} from "../../../models/ICalendar";

interface CalendarProps {
    calendar: ICalendar,
    calendarState: ICalendarState
}

const Calendar: FC<CalendarProps> = ({calendar, calendarState}) => {
    let calendarArray = calendar.getCalendar(calendarState.month, calendarState.year).map(item=> {
        return item.map(weekDay => {
            return (
                <Grid item sm={1.7} key={Math.random()}>
                    <CalendarItem
                        sx={{backgroundColor: weekDay.isActive ? '#484848' : '#343434'}}
                    >
                        <Box
                            width={30}
                            height={30}
                            sx={{backgroundColor: weekDay.isToday ? 'red' : 'transparent', padding: '0 2px', borderRadius: '100%'}}
                            display={'flex'}
                            justifyContent={'center'}
                            alignItems={'center'}
                        >
                            <Typography variant={'body1'}>
                                {weekDay.day}
                            </Typography>
                        </Box>
                    </CalendarItem>
                </Grid>
            )
        })
    })
    return (
        <Grid container justifyContent={'space-between'} rowGap={'2px'}>
            {calendarArray}
        </Grid>
    );
};

export default Calendar;