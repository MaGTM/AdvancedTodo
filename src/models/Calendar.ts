import {ICalendar, IWeek} from "./ICalendar";


export class Calendar implements ICalendar {
    readonly currentYear: number
    readonly currentMonth: number
    readonly currentDate: number
    readonly currentWeekDay: number
    private calendar: IWeek[][] = []

    constructor() {
        let date = new Date()
        this.currentYear = date.getFullYear()
        this.currentMonth = date.getMonth()
        this.currentDate = date.getDate()
        this.currentWeekDay = date.getDay()
    }

    private daysInMonth(month: number) {
        let date = new Date()
        date.setMonth(month)
        return 33 - new Date(date.getFullYear(), date.getMonth(), 33).getDate();
    }

    private getNormalDay(day: number) {
        if (day === 0) {
            return 6
        }
        return day - 1
    }

    getCalendar(month: number, year: number) { // Getting Calendar Days
        let weekArray: IWeek[] = []
        this.calendar = []
        let startCurrentMonth = new Date()
        startCurrentMonth.setMonth(month)
        startCurrentMonth.setFullYear(year)
        startCurrentMonth.setDate(1)
        let currentMonthFirstDayOfWeek = this.getNormalDay(startCurrentMonth.getDay())

        for (let i = this.daysInMonth(startCurrentMonth.getMonth() - 1) - currentMonthFirstDayOfWeek + 1;
             i <= this.daysInMonth(startCurrentMonth.getMonth() - 1);
             i++) {
            weekArray.push({day: i, isActive: false})
        }

        for (let i = currentMonthFirstDayOfWeek; i < 7; i++) {
            weekArray.push({
                day: i - currentMonthFirstDayOfWeek + 1,
                isActive: true,
                isToday: (i - currentMonthFirstDayOfWeek + 1) === this.currentDate && month === this.currentMonth && year === this.currentYear
            })
        }

        this.calendar.push(weekArray)
        let nextMonth = false
        for (let i = 0; i < 5; i++) {
            weekArray = []
            for (let j = this.calendar[this.calendar.length - 1][6].day + 1; j < this.calendar[this.calendar.length - 1][6].day + 8; j++) {
                if (j > this.daysInMonth(startCurrentMonth.getMonth())) {
                    weekArray.push({day: j - this.daysInMonth(startCurrentMonth.getMonth()), isActive: false})
                    nextMonth = true
                    continue
                }
                if(nextMonth) {
                    weekArray.push({day: j, isActive: false})
                    continue
                }
                weekArray.push({day: j, isActive: true, isToday: this.currentDate === j && month === this.currentMonth && year === this.currentYear})
            }
            this.calendar.push(weekArray)
        }


        return this.calendar
    }
}