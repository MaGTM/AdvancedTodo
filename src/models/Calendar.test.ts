import {Calendar} from "./Calendar";

const calendar = new Calendar()
const date = new Date()

describe('Calendar Model', () => {
    test('Month 1', () => {
        expect(calendar.getCalendar(date.getMonth(), date.getFullYear())).toMatchSnapshot()
    })
    test('Month 2', () => {
        expect(calendar.getCalendar(date.getMonth()-1, date.getFullYear())).toMatchSnapshot()
    })
    test('Year 1', () => {
        expect(calendar.getCalendar(date.getMonth(), date.getFullYear()+1)).toMatchSnapshot()
    })
    test('Year 2', () => {
        expect(calendar.getCalendar(date.getMonth(), date.getFullYear()-1)).toMatchSnapshot()
    })
})
