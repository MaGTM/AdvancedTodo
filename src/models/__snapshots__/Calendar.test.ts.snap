// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Model Month 1 1`] = `
Array [
  Array [
    Object {
      "day": 28,
      "isActive": false,
    },
    Object {
      "day": 29,
      "isActive": false,
    },
    Object {
      "day": 30,
      "isActive": false,
    },
    Object {
      "day": 1,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 2,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 3,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 4,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 5,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 6,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 7,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 8,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 9,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 10,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 11,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 12,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 13,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 14,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 15,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 16,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 17,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 18,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 19,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 20,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 21,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 22,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 23,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 24,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 25,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 26,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 27,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 28,
      "isActive": true,
      "isToday": true,
    },
    Object {
      "day": 29,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 30,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 31,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 1,
      "isActive": false,
    },
  ],
  Array [
    Object {
      "day": 2,
      "isActive": false,
    },
    Object {
      "day": 3,
      "isActive": false,
    },
    Object {
      "day": 4,
      "isActive": false,
    },
    Object {
      "day": 5,
      "isActive": false,
    },
    Object {
      "day": 6,
      "isActive": false,
    },
    Object {
      "day": 7,
      "isActive": false,
    },
    Object {
      "day": 8,
      "isActive": false,
    },
  ],
]
`;

exports[`Calendar Model Month 2 1`] = `
Array [
  Array [
    Object {
      "day": 31,
      "isActive": false,
    },
    Object {
      "day": 1,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 2,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 3,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 4,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 5,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 6,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 7,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 8,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 9,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 10,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 11,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 12,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 13,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 14,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 15,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 16,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 17,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 18,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 19,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 20,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 21,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 22,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 23,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 24,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 25,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 26,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 27,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 28,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 29,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 30,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 1,
      "isActive": false,
    },
    Object {
      "day": 2,
      "isActive": false,
    },
    Object {
      "day": 3,
      "isActive": false,
    },
    Object {
      "day": 4,
      "isActive": false,
    },
  ],
  Array [
    Object {
      "day": 5,
      "isActive": false,
    },
    Object {
      "day": 6,
      "isActive": false,
    },
    Object {
      "day": 7,
      "isActive": false,
    },
    Object {
      "day": 8,
      "isActive": false,
    },
    Object {
      "day": 9,
      "isActive": false,
    },
    Object {
      "day": 10,
      "isActive": false,
    },
    Object {
      "day": 11,
      "isActive": false,
    },
  ],
]
`;

exports[`Calendar Model Year 1 1`] = `
Array [
  Array [
    Object {
      "day": 27,
      "isActive": false,
    },
    Object {
      "day": 28,
      "isActive": false,
    },
    Object {
      "day": 29,
      "isActive": false,
    },
    Object {
      "day": 30,
      "isActive": false,
    },
    Object {
      "day": 1,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 2,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 3,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 4,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 5,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 6,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 7,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 8,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 9,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 10,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 11,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 12,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 13,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 14,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 15,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 16,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 17,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 18,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 19,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 20,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 21,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 22,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 23,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 24,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 25,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 26,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 27,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 28,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 29,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 30,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 31,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 1,
      "isActive": false,
    },
    Object {
      "day": 2,
      "isActive": false,
    },
    Object {
      "day": 3,
      "isActive": false,
    },
    Object {
      "day": 4,
      "isActive": false,
    },
    Object {
      "day": 5,
      "isActive": false,
    },
    Object {
      "day": 6,
      "isActive": false,
    },
    Object {
      "day": 7,
      "isActive": false,
    },
  ],
]
`;

exports[`Calendar Model Year 2 1`] = `
Array [
  Array [
    Object {
      "day": 29,
      "isActive": false,
    },
    Object {
      "day": 30,
      "isActive": false,
    },
    Object {
      "day": 1,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 2,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 3,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 4,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 5,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 6,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 7,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 8,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 9,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 10,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 11,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 12,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 13,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 14,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 15,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 16,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 17,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 18,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 19,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 20,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 21,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 22,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 23,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 24,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 25,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 26,
      "isActive": true,
      "isToday": false,
    },
  ],
  Array [
    Object {
      "day": 27,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 28,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 29,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 30,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 31,
      "isActive": true,
      "isToday": false,
    },
    Object {
      "day": 1,
      "isActive": false,
    },
    Object {
      "day": 2,
      "isActive": false,
    },
  ],
  Array [
    Object {
      "day": 3,
      "isActive": false,
    },
    Object {
      "day": 4,
      "isActive": false,
    },
    Object {
      "day": 5,
      "isActive": false,
    },
    Object {
      "day": 6,
      "isActive": false,
    },
    Object {
      "day": 7,
      "isActive": false,
    },
    Object {
      "day": 8,
      "isActive": false,
    },
    Object {
      "day": 9,
      "isActive": false,
    },
  ],
]
`;
